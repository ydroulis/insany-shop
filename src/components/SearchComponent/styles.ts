'use client'

import styled from 'styled-components'
import { CiSearch } from "react-icons/ci";

export const Container = styled.div`
    position: relative;
    width: 352px;
    height: 42px;

    border-radius: 8px;
`

export const SearchInput = styled.input`
    width: 100%;
    height: 42px;
    border: none;

    border-radius: 8px;
    background-color: #F3F7FF;

    padding: 0rem .75rem;

    color: #737380;

    &::placeholder {
        color: #737380;
        font-family: 'Inter', sans-serif;
    }

    &::-webkit-search-decoration,
    &::-webkit-search-cancel-button,
    &::-webkit-search-results-button,
    &::-webkit-search-results-decoration {
        display: none;
    }

    &::-webkit-search-cancel-button {
        -webkit-appearance: none;
    }
`

export const SearchIcon = styled(CiSearch)`
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 9px;
    color: #5D5D6D;
`

export const SrOnly = styled.span`
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    border: 0;
`

export const Suggestions = styled.ul`
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  background: #fff;
  border: 1px solid #ddd;
  border-radius: 8px;
  margin-top: 4px;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
  list-style: none;
  padding: 4px 0;
  max-height: 220px;
  overflow-y: auto;
  z-index: 10;
`;

export const SuggestionItem = styled.li`
  padding: 8px 12px;
  cursor: pointer;
  font-size: 14px;

  &:hover {
    background: #f5f5f5;
  }
`;

export const SearchWrapper = styled.div`
  position: relative;
  width: 100%;
`;

export const ClearButton = styled.button`
  position: absolute;
  right: 36px; /* antes do Ã­cone de busca */
  top: 50%;
  transform: translateY(-50%);
  background: transparent;
  border: none;
  font-size: 18px;
  cursor: pointer;
  color: #666;

  &:hover {
    color: #000;
  }
`;