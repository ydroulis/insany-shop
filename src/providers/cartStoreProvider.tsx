'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CartStore, createCartStore } from '../stores/cartStore'

export type CartStoreApi = ReturnType<typeof createCartStore>

export const CartStoreContext = createContext<CartStoreApi | undefined>(
    undefined,
)

export interface CartStoreProviderProps {
    children: ReactNode
}

export const CartStoreProvider = ({
    children,
}: CartStoreProviderProps) => {
    const storeRef = useRef<CartStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCartStore()
    }

    return (
        <CartStoreContext.Provider value={storeRef.current}>
            {children}
        </CartStoreContext.Provider>
    )
}

export const useCartStore = <T,>(
    selector: (store: CartStore) => T,
): T => {
    const cartStoreContext = useContext(CartStoreContext)

    if (!cartStoreContext) {
        throw new Error(`useStore must be used within CartStoreProvider`)
    }

    return useStore(cartStoreContext, selector)
}
