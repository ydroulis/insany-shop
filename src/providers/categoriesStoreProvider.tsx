'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CategoriesStore, createCategoriesStore } from '../stores/categoriesStore'

export type CategoriesStoreApi = ReturnType<typeof createCategoriesStore>

export const CategoriesStoreContext = createContext<CategoriesStoreApi | undefined>(
    undefined,
)

export interface CategoriesStoreProviderProps {
    children: ReactNode
}

export const CategoriesStoreProvider = ({
    children,
}: CategoriesStoreProviderProps) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCategoriesStore()
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    )
}

export const useCategoriesStore = <T,>(
    selector: (store: CategoriesStore) => T,
): T => {
    const categoriesStoreContext = useContext(CategoriesStoreContext)

    if (!categoriesStoreContext) {
        throw new Error(`useStore must be used within CategoriesStoreProvider`)
    }

    return useStore(categoriesStoreContext, selector)
}
