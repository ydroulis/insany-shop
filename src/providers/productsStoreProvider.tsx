'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type ProductsStore, createProductsStore } from '../stores/productsStore'

export type ProductsStoreApi = ReturnType<typeof createProductsStore>

export const ProductsStoreContext = createContext<ProductsStoreApi | undefined>(
    undefined,
)

export interface ProductsStoreProviderProps {
    children: ReactNode
}

export const ProductsStoreProvider = ({
    children,
}: ProductsStoreProviderProps) => {
    const storeRef = useRef<ProductsStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createProductsStore()
    }

    return (
        <ProductsStoreContext.Provider value={storeRef.current}>
            {children}
        </ProductsStoreContext.Provider>
    )
}

export const useProductsStore = <T,>(
    selector: (store: ProductsStore) => T,
): T => {
    const productsStoreContext = useContext(ProductsStoreContext)

    if (!productsStoreContext) {
        throw new Error(`useStore must be used within ProductsStoreProvider`)
    }

    return useStore(productsStoreContext, selector)
}
