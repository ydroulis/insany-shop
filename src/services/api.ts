import axios from "axios";

export const api = axios.create({
    baseURL: "https://api.insany.co/api",
    headers: {
        "Content-Type": "application/json",
    },
});