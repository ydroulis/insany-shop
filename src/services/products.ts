import { Product, ProductsResponse } from "@/types/Products";
import { api } from "./api";

export const getProducts = async (
    page?: number,
    limit?: number,
    category?: string
): Promise<ProductsResponse> => {
    const params: Record<string, string | number> = {};
    if (page) params.page = page;
    if (limit) params.limit = limit;
    if (category) params.category = category;

    const response = await api.get<ProductsResponse>("/products", { params });
    return response.data;
};

export const getProductById = async (id: number): Promise<Product> => {
    const response = await api.get<{ product: Product }>(`/products/${id}`);
    return response.data.product;
};
