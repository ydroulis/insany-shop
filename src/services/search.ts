import { api } from "./api";

export interface SearchResponse {
    suggestions: string[];
    query: string;
}

export const searchProducts = async (query: string): Promise<SearchResponse> => {
    const response = await api.get<SearchResponse>("/search", { params: { q: query } });
    return response.data;
};
