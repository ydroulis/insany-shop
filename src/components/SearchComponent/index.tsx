import React, { useEffect, useState } from 'react';
import * as S from './styles';
import { searchProducts, SearchResponse } from '../../services/search';
import { useProductsStore } from '../../providers/productsStoreProvider';
import { getProducts } from '../../services/products';
import { Products } from '@/types/Products';

const SearchComponent: React.FC = () => {
    const [inputValue, setInputValue] = useState('');
    const [results, setResults] = useState<SearchResponse | null>(null);
    const [loading, setLoading] = useState(false);
    const [isOpen, setIsOpen] = useState(false);
    const [allProducts, setAllProducts] = useState<Products>([]);

    const { setProducts } = useProductsStore((state) => state);

    useEffect(() => {
        if (!inputValue.trim()) {
            setResults(null);
            setIsOpen(false);
            return;
        }

        const fetchAllProducts = async () => {
            const res = await getProducts(1, 100);
            setAllProducts(res.products);
        };
        fetchAllProducts();

        setLoading(true);
        searchProducts(inputValue)
            .then((response) => {
                setResults(response);
            })
            .finally(() => {
                setLoading(false);
            });
    }, [inputValue, setProducts]);

    useEffect(() => {
        if (inputValue.trim() === '') {

            const fetchProducts = async () => {
                const response = await getProducts();
                setProducts(response.products);
            }
            fetchProducts();
        }
    }, [inputValue, setProducts]);

    const handleClick = (item: string) => {
        const searchedProducts = allProducts.filter((product) =>
            product.name.toLowerCase().includes(item.toLowerCase())
        );
        setInputValue(item);
        setProducts(searchedProducts);
        setIsOpen(false);
    };

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const value = e.target.value;
        setInputValue(value);

        if (value.trim() === '') {
            setIsOpen(false);
            setResults(null);
        } else {
            setIsOpen(true);
        }
    };

    const handleClear = () => {
        setInputValue('');
        setResults(null);
    };

    return (
        <S.Container role="search" aria-label="Pesquisar no site">
            <label htmlFor="site-search" className="sr-only">
                <S.SrOnly>Procurar produtos</S.SrOnly>
            </label>

            <S.SearchWrapper>
                <S.SearchInput
                    value={inputValue}
                    onChange={handleChange}
                    id="site-search"
                    type="search"
                    placeholder="Procurando algo específico"
                    aria-label="Campo de busca"
                />

                <S.SearchIcon
                    data-testid="search-icon"
                    size={24}
                    aria-hidden={true}
                    focusable="false"
                />

                {inputValue && (
                    <S.ClearButton onClick={handleClear} aria-label="Limpar busca">
                        ×
                    </S.ClearButton>
                )}

                {loading && <p>Carregando...</p>}

                {isOpen && results && results.suggestions.length > 0 && (
                    <S.Suggestions role="listbox" aria-label="Sugestões de busca">
                        {!loading &&
                            results.suggestions.map((s, i) => (
                                <S.SuggestionItem
                                    key={i}
                                    role="option"
                                    aria-selected={false}
                                    tabIndex={0}
                                    onClick={() => handleClick(s)}
                                    onKeyDown={(e) => {
                                        if (e.key === 'Enter') {
                                            e.preventDefault();
                                            handleClick(s);
                                        }
                                    }}
                                >
                                    {s}
                                </S.SuggestionItem>
                            ))}
                    </S.Suggestions>
                )}
            </S.SearchWrapper>
        </S.Container>
    );
};

export default SearchComponent;
