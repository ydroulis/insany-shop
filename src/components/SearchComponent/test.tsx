import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import SearchComponent from './';
import { searchProducts } from '../../services/search';
import { getProducts } from '../../services/products';
import { useProductsStore } from '../../providers/productsStoreProvider';

jest.mock('../../services/search', () => ({
    searchProducts: jest.fn().mockResolvedValue({ suggestions: ['Produto A'], query: 'Produto' }),
}));

jest.mock('../../services/products', () => ({
    getProducts: jest.fn().mockResolvedValue({ products: [{ name: 'Produto A' }] }),
}));

jest.mock('../../providers/productsStoreProvider', () => ({
    useProductsStore: jest.fn(),
}));

describe('<SearchComponent />', () => {
    const mockSetProducts = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useProductsStore as jest.Mock).mockReturnValue({
            setProducts: mockSetProducts,
        });
    });

    it('should render the search input, icon, and clear button', () => {
        render(<SearchComponent />);

        const container = screen.getByRole('search', { name: /pesquisar no site/i });
        expect(container).toBeInTheDocument();

        const input = screen.getByRole('searchbox', { name: /campo de busca/i });
        expect(input).toBeInTheDocument();
        expect(input).toHaveAttribute('placeholder', 'Procurando algo especÃ­fico?');

        const icon = screen.getByTestId('search-icon');
        expect(icon).toBeInTheDocument();
    });

    it('should open suggestions when typing and allow selection by click', async () => {
        (searchProducts as jest.Mock).mockResolvedValue({
            suggestions: ['Produto A', 'Produto B'],
            query: 'Pro',
        });
        (getProducts as jest.Mock).mockResolvedValue({ products: [{ name: 'Produto A' }, { name: 'Produto B' }] });

        render(<SearchComponent />);

        const input = screen.getByRole('searchbox', { name: /campo de busca/i });
        fireEvent.change(input, { target: { value: 'Pro' } });

        await waitFor(() => expect(searchProducts).toHaveBeenCalledWith('Pro'));

        const suggestionA = await screen.findByText('Produto A');
        expect(suggestionA).toBeInTheDocument();

        fireEvent.click(suggestionA);
        expect(mockSetProducts).toHaveBeenCalledWith([{ name: 'Produto A' }]);
        expect(input).toHaveValue('Produto A');
    });

    it('should allow selection of suggestion with Enter key', async () => {
        (searchProducts as jest.Mock).mockResolvedValue({
            suggestions: ['Produto C'],
            query: 'Pro',
        });
        (getProducts as jest.Mock).mockResolvedValue({ products: [{ name: 'Produto C' }] });

        render(<SearchComponent />);
        const input = screen.getByRole('searchbox', { name: /campo de busca/i });

        fireEvent.change(input, { target: { value: 'Pro' } });

        const suggestionC = await screen.findByText('Produto C');
        suggestionC.focus();

        fireEvent.keyDown(suggestionC, { key: 'Enter', code: 'Enter' });

        expect(mockSetProducts).toHaveBeenCalledWith([{ name: 'Produto C' }]);
        expect(input).toHaveValue('Produto C');
    });

    it('should clear input when clicking the clear button', async () => {
        render(<SearchComponent />);

        const input = screen.getByRole('searchbox', { name: /campo de busca/i });
        fireEvent.change(input, { target: { value: 'Teste' } });
        expect(input).toHaveValue('Teste');

        const clearButton = screen.getByRole('button', { name: /limpar busca/i });
        fireEvent.click(clearButton);

        expect(input).toHaveValue('');
    });

    it('should match snapshot', () => {
        const { container } = render(<SearchComponent />);
        expect(container).toMatchSnapshot();
    });
});
