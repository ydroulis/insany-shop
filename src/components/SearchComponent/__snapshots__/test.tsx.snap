// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<SearchComponent /> should match snapshot 1`] = `
<div>
  <div
    aria-label="Pesquisar no site"
    class="sc-bRKDuR gmhrnS"
    role="search"
  >
    <label
      class="sr-only"
      for="site-search"
    >
      <span
        class="sc-ggWZvA garkgY"
      >
        Procurar produtos
      </span>
    </label>
    <div
      class="sc-hjsuWn gZDtLR"
    >
      <input
        aria-label="Campo de busca"
        class="sc-hvigdm jqpAwf"
        id="site-search"
        placeholder="Procurando algo especÃ­fico?"
        type="search"
        value=""
      />
      <svg
        aria-hidden="true"
        class="sc-fhHczv gkiTve"
        data-testid="search-icon"
        fill="currentColor"
        focusable="false"
        height="24"
        id="search"
        stroke="currentColor"
        stroke-width="0"
        version="1.1"
        viewBox="0 0 24 24"
        width="24"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <path
            d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
          />
        </g>
      </svg>
    </div>
  </div>
</div>
`;
